### ISSUE TITLE
Performance regression using FileField and Django 3.1

### ISSUE THREAD

**Issue by LincolnPuzey** [romgar/django-dirtyfields#165]
I've met problem with using `dirtyfields` for `FileField`. Previous Django version used simpler implementation of `__getstate__`, and 3.1 add `instance` to state, so `deepcopy` also copies instance, which is much slower and consume more memory.

Django changes: https://github.com/django/django/commit/f600e3fad6e92d9fe1ad8b351dc8446415f24345
deepcopy: https://github.com/romgar/django-dirtyfields/blob/develop/src/dirtyfields/dirtyfields.py#L95

_Originally posted by @ron8mcr in https://github.com/romgar/django-dirtyfields/issues/153#issuecomment-678922807_

**Comment by dryan** [comment]
I'm getting errors because of this it seems. `cannot pickle '_io.BufferedReader' object` on instances with FileFields. Full traceback below:

```
Traceback (most recent call last):
  File "/Users/dryan/Sites/prideraiser.org/accounts/tests/test_models.py", line 95, in test_get_last_pledge_notification_preference
    PledgeFactory.create(user=user)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/base.py", line 564, in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/django.py", line 141, in _generate
    return super(DjangoModelFactory, cls)._generate(strategy, params)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/base.py", line 501, in _generate
    return step.build()
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/builder.py", line 272, in build
    step.resolve(pre)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/builder.py", line 221, in resolve
    self.attributes[field_name] = getattr(self.stub, field_name)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/builder.py", line 372, in __getattr__
    value = value.evaluate(
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/declarations.py", line 321, in evaluate
    return self.generate(step, defaults)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/declarations.py", line 411, in generate
    return step.recurse(subfactory, params, force_sequence=force_sequence)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/builder.py", line 233, in recurse
    return builder.build(parent_step=self, force_sequence=force_sequence)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/builder.py", line 276, in build
    instance = self.factory_meta.instantiate(
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/base.py", line 315, in instantiate
    return self.factory._create(model, *args, **kwargs)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/factory/django.py", line 185, in _create
    return manager.create(*args, **kwargs)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/django/db/models/query.py", line 445, in create
    obj = self.model(**kwargs)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/dirtyfields/dirtyfields.py", line 38, in __init__
    reset_state(sender=self.__class__, instance=self)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/dirtyfields/dirtyfields.py", line 155, in reset_state
    new_state = instance._as_dict(check_relationship=True)
  File "/Users/dryan/.pyenv/versions/3.9.1/envs/prideraiser/lib/python3.9/site-packages/dirtyfields/dirtyfields.py", line 94, in _as_dict
    all_field[field.name] = deepcopy(field_value)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/dryan/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 161, in deepcopy
    rv = reductor(4)
TypeError: cannot pickle '_io.BufferedReader' object
```

**Comment by LincolnPuzey** [comment]
@dryan what value are you assigning to the model field? It appears it can't be deepcopied/pickled.

**Comment by dryan** [comment]
It's an uploaded image.

**Comment by LincolnPuzey** [comment]
@dryan Can you raise this in a new issue and include your code or a minimal example. something resulting in a stacktrace is not a performance regression.

**Comment by dryan** [comment]
@LincolnPuzey seems to have been an issue with how Factory Boy was generating ImageField values that they've fixed.

**Comment by himynameistimli** [comment]
I seem to get a similar issue on my end. For my situation, I'm copying an image field (to a preview image field) when the image is first saved to my model via a DRF serializer.
 
When upgrading to Django 3.1 (from 3.0) I get an error that seems to be related to the `deepcopy` function. When I remove the `DirtyFieldsMixin` mixin from my model, the error disappears.

```bash
Traceback (most recent call last):
  File "/[my_env]/lib/python3.7/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/[my_env]/lib/python3.7/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/[my_env]/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/mixins.py", line 19, in create
    self.perform_create(serializer)
  File "/[my_app]/viewsets.py", line 323, in perform_create
    serializer.save(
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/serializers.py", line 205, in save
    self.instance = self.create(validated_data)
  File "/[my_app]/serializers.py", line 194, in create
    return super().create(validated_data)
  File "/[my_env]/lib/python3.7/site-packages/rest_framework/serializers.py", line 939, in create
    instance = ModelClass._default_manager.create(**validated_data)
  File "/[my_env]/lib/python3.7/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/[my_env]/lib/python3.7/site-packages/django/db/models/query.py", line 445, in create
    obj = self.model(**kwargs)
  File "/[my_env]/lib/python3.7/site-packages/dirtyfields/dirtyfields.py", line 38, in __init__
    reset_state(sender=self.__class__, instance=self)
  File "/[my_env]/lib/python3.7/site-packages/dirtyfields/dirtyfields.py", line 155, in reset_state
    new_state = instance._as_dict(check_relationship=True)
  File "/[my_env]/lib/python3.7/site-packages/dirtyfields/dirtyfields.py", line 94, in _as_dict
    all_field[field.name] = deepcopy(field_value)
  File "/[my_env]/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/[my_env]/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/[my_env]/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/[my_env]/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/[my_env]/lib/python3.7/copy.py", line 169, in deepcopy
    rv = reductor(4)
  File "/[my_env]/lib/python3.7/site-packages/PIL/Image.py", line 696, in __getstate__
    return [self.info, self.mode, self.size, self.getpalette(), self.tobytes()]
  File "/[my_env]/lib/python3.7/site-packages/PIL/Image.py", line 1350, in getpalette
    self.load()
  File "/[my_env]/lib/python3.7/site-packages/PIL/ImageFile.py", line 183, in load
    seek = self.fp.seek
AttributeError: 'NoneType' object has no attribute 'seek'
Internal Server Error: /api/v1/events/ecdf0af1-2aa0-48e4-bd6d-3a52520d25af/screenshots/
Traceback (most recent call last):
  File "/[my_env]/lib/python3.7/site-packages/PIL/ImageFile.py", line 180, in load
    seek = self.load_seek
AttributeError: 'PngImageFile' object has no attribute 'load_seek'
```

**Comment by LincolnPuzey** [comment]
@himynameistimli might be a bug in `PIL`? since the error is coming from inside `PIL/Image.py", line 696, in __getstate__`

Otherwise django-dirtyfields is deepcopying something it shouldn't.

**Comment by LincolnPuzey** [comment]
For FileField and ImageField,

> "All that will be stored in your database is a path to the file"

My proposal for resolving the performance regression and reducing the chance of errors like the above is for django-dirtyfields to only look at / care about the filepath string saved to the database and ignore the file object / the actual saved files.

This would mean:
1. these fields would be represented as their filepath strings in `_as_dict` and `get_dirtyfields` (potenially breaking change)
2. these fields would be considered dirty iff the filepath changes. (current behavior is not well defined). 
3. changing content of the file without changing filepath -> not dirty
4. changing filepath but keeping content the same -> dirty

i.e. dirtyfields will never need to look at the file object in storage, and the memory usage for these fields is just the filepath string.

I think most django apps probably just generate a filepath for a file when its uploaded, then don't change it in any way until replaced with a new uploaded file. This works well with this well-defined definition of "dirty"

Also Django's `FieldFile` class has the following comment:

> Older code may be expecting FileField values to be simple strings.

So it may be that my proposal is how dirtyfields worked for older versions of django.

**Comment by AlbericTrancart** [comment]
Hello! Any update on this issue? We are encountering this on a FileField, it doubles the time of our test pipeline. Is there any way to ignore this field from the dirty fields method?

**Comment by LincolnPuzey** [comment]
Fixed in #203
